clang-tools-extra/clangd/refactor/tweaks/AbbreviateFunctionTemplate.cpp

```diff
@@ -0,0 +1,310 @@
+//===-- AbbreviateFunctionTemplate.cpp ---------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "FindTarget.h"
+#include "SourceCode.h"
+#include "XRefs.h"
+#include "refactor/Tweak.h"
+#include "support/Logger.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/ExprConcepts.h"
+#include "clang/Tooling/Core/Replacement.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include <numeric>
+
+namespace clang {
+namespace clangd {
+namespace {
+/// Converts a function template to its abbreviated form using auto parameters.
+/// Before:
+///     template <std::integral T>
+///     auto foo(T param) { }
+///          ^^^^^^^^^^^
+/// After:
+///     auto foo(std::integral auto param) { }
+class AbbreviateFunctionTemplate : public Tweak {
+public:
+  const char *id() const final;
```

Comment:
I'm not sure whether the comment behind the [documentation](https://clang.llvm.org/extra/doxygen/classclang_1_1clangd_1_1Tweak.html) ` // defined by REGISTER_TWEAK` is meant to literally being copied or just as a hint for the implementer. I was a bit confused at first to not find a definition of the function.

This declaration as is breaks the consistency of using trailing-return types.

Reply:
`const char *id() const final;` is copied like this in each tweak.

@jeremystucki  I think we need to remoe the trailing-return types as they are used only in our refactorings.

Reply:
> @jeremystucki I think we need to remoe the trailing-return types as they are used only in our refactorings.

As with the previous refactoring I suggest submitting the patch as-is and only changing it if they ask us to.